src/editable_tree/mod.rs

```diff
@@ -190,56 +322,113 @@ impl<'arena, Node: Ast<'arena>> DAG<'arena, Node> {
         self.history_index = self.root_history.len() - 1;
     }
 
-    /// Updates the internal state so that the tree now contains `new_node` in the position of the
-    /// `cursor`.
-    pub fn replace_cursor(&mut self, new_node: Node) {
+    /// Replaces the current cursor with a node represented by `c`
+    pub fn replace_cursor(&mut self, c: char) -> EditResult {
+        /* CHECK THE VALIDITY OF ARGUMENTS */
+
+        // Cache the node under the cursor, since finding the cursor involves non-trivial amounts
+        // of work
+        let cursor = self.cursor();
+        // Short circuit if the char to insert couldn't correspond to a valid child
+        if !cursor.is_replace_char(c) {
+            return EditResult::Err(EditErr::CannotReplace(c));
+        }
+
+        /* PERFORM THE ACTION */
+
+        // We require that `c` corresponds to a valid node (because `is_replace_char` returned
+        // `true`), so we can unwrap
+        let new_node = cursor.from_char(c).unwrap();
         // Generate a vec of pointers to the nodes that we will have to clone.  We have to store
         // this as a vec because the iterator that produces them (cursor_path::NodeIter) can only
         // yield values from the root downwards, whereas we need the nodes in the opposite order.
         let mut nodes_to_clone: Vec<_> = self.current_cursor_path.node_iter(self.root()).collect();
         // The last value of nodes_to_clone is the node under the cursor, which we do not need to
         // clone, so we pop that reference.
         assert!(nodes_to_clone.pop().is_some());
-        self.finish_edit(&nodes_to_clone, new_node);
+
+        /* FINISH THE EDIT AND RETURN SUCCESS */
+
+        // Store the new_node's display name before it's consumed by `finish_edit`
+        let new_node_name = new_node.display_name();
+        self.finish_edit(&nodes_to_clone, 0, new_node);
+        EditResult::Ok(EditSuccess::Replace {
+            c,
+            name: new_node_name,
+        })
     }
 
-    /// Updates the internal state so that the tree now contains `new_node` inserted as the first
+    /// Updates the internal state so that the tree now contains `new_node` inserted as the last
     /// child of the selected node.  Also moves the cursor so that the new node is selected.
-    pub fn insert_child(&mut self, new_node: Node) -> Result<(), Node::InsertError> {
+    pub fn insert_child(&mut self, c: char) -> EditResult {
+        /* CHECK THE VALIDITY OF ARGUMENTS */
+
+        // Cache the node under the cursor, since finding the cursor involves non-trivial amounts
+        // of work
+        let cursor = self.cursor();
+        // Short circuit if `c` couldn't be a valid child of the cursor
+        if !cursor.is_insert_char(c) {
+            return EditResult::Err(EditErr::CannotInsertInto {
+                c,
+                parent_name: cursor.display_name(),
```

Comment:
I don't remember if `display_name` allocates. If it does `ok_or_else()` should be used instead of `ok_or()` here.

Reply:
I've made all the `ok_or`s return `EditErr::CharNotANode(char)` which doesn't allocate - so this is fine.